shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -27,6 +28,54 @@
 
 NSNotificationName const FlutterViewControllerWillDealloc = @"FlutterViewControllerWillDealloc";
 
+namespace {
+/// Provides a mechanism to pause a TaskRunner.
+class TaskRunnerPause {
+ public:
+  /// Create a `TaskRunnerPause`.
+  /// @param[in]  task_runner    The `TaskRunner` we will pause.
+  TaskRunnerPause(fml::RefPtr<fml::TaskRunner> task_runner) : task_runner_(task_runner) {}
+
+  ~TaskRunnerPause() {
+    // Flush task runner to make sure it isn't in the middle of wrapping up resuming.
+    fml::AutoResetWaitableEvent latch;
+    task_runner_->PostTask([&] { latch.Signal(); });
+    latch.Wait();
```

Comment:
Wont' this deadlock if we're in a `Pause` that we didn't `resume`?

For example, if the VC goes out of scope after calling `Pause` but before the app is fully backgrounded for some reason.

Reply:
Yep, by design.  The only way that could happen is if an app asks for a background task and that background task trashes the VC.  That may be something someone might do to try to alleviate memory usage.  I'm not quite sure they could recover the VC when the app comes back.

Reply:
My point is, that might be something someone does.  But is it something we support otherwise and make sense?

Reply:
The problem is it will be really hard for an end user to figure out what's going on in that case right?  There's a path here where a user (or a plugin, or some weirdness in an existing app) can cause a deadlock here that wouldn't have existed previously.

Reply:
Ok, addressed.

Reply:
If we just call `Resume`, why do we want to drain the task runner? Couldn't someone still get behind us once this latch gives up, and couldn't it still be a spot for a deadlock to occur?

Reply:
We need to drain the task runner because it could still be using the mutex momentarily as part of using the condition variable and we are about to delete the mutex.

> Couldn't someone still get behind us once this latch gives up, and couldn't it still be a spot for a deadlock to occur?

Nope.  I mean someone would have to pause the thread from a different thread at this exact moment, then wait for this thread for some reason to complete something.  No one is doing that.